rfd/0000-prototype-job-worker.md

```diff
@@ -0,0 +1,203 @@
+---
+author: Bill Rich (bill@dangeri.ch)
+state: draft
+---
+ 
+# RFD 0000 - Prototype Job Worker Service 
+
+## Required Approvers
+
+- Engineering: @rosstimothy || @zmb3, Marek Smoliński || @smallinsky, Anton Miniailo || @AntonAM
+
+## What
+
+This document presents the design for a prototype job worker service that
+manages and executes arbitrary Linux processes. It details the components,
+functionality, and architectural considerations of the service, including the
+library, API, and client-side interactions. The service is designed to provide
+APIs for starting, stopping, querying the status of jobs, and streaming the
+output of running jobs.
+
+## Why
+
+The purpose of creating this prototype is to establish a foundation for a
+scalable, secure, and efficient way to handle process management on Linux
+systems. This service aims to streamline development operations and ensure
+robust process isolation and resource control, enhancing the security and
+performance of running jobs. This prototype also serves as a critical step
+towards evaluating the practical aspects of process management in controlled
+environments, focusing on rapid development and deployment with minimal code
+complexity.
+
+## Details
+
+### Components
+#### Library
+- **Functionality**: The library will include methods to start, stop, query
+  status, and get the output of jobs. It will handle streaming output from the
+  inception of process execution. 
+- **Concurrency**: Support for multiple concurrent clients will be implemented.
+  Resource Control: Utilize cgroups for CPU, memory, and disk I/O control per job.
+  Resource Isolation: Implement PID, mount, and network namespaces to ensure job isolation.
+#### API
+- **Technology**: gRPC will be used due to its high performance and suitability
+  for streaming data.
+- **Methods**:
+  - `StartJob`: Starts a new job.
+  - `StopJob`: Terminates an existing job.
+  - `GetJobStatus`: Returns the current status of a job.
+  - `StreamOutput`: Streams the output of a running job.
+- **Security**: Utilize mTLS for authentication with a strong set of cipher
+  suites and certificate configurations. 
+#### Client
+- **Interface**: A CLI that can interact with the worker service to start,
+  stop, get status, and stream the output of jobs. 
+- **Features**: The CLI will communicate with the server using the gRPC API,
+  handling all server responses and streaming outputs effectively.
+
+### Detailed Design
+#### Streaming Approach
+- **Implementation**: Utilize gRPC streaming to send output from the worker to
+  the client in real-time. Buffering strategies will be applied to handle
+  high-frequency output efficiently.
+#### TLS Setup
+- **TLS Version**: TLS 1.3 will be used due to its enhanced security features.
+- **Cipher Suites**: Use strong cipher suites such as TLS_AES_256_GCM_SHA384 to
+  ensure secure communication.
```

Comment:
I was pretty sure it was possible. I'll need to double check, but I thought this would work:
```
	config := &tls.Config{
		Certificates: []tls.Certificate{cert},
		ClientAuth:   tls.RequireAndVerifyClientCert,
		ClientCAs:    caCertPool,
		MinVersion:   tls.VersionTLS13,
		CipherSuites: []uint16{tls.TLS_AES_256_GCM_SHA384},
	}
```